vllm/v1/sample/sampler.py

```diff
@@ -227,3 +237,17 @@ def apply_logits_bias(
                 for token_id, bias in logit_bias.items():
                     logits[i, token_id] += bias
         return logits
+
+    def apply_allowed_token_ids(
+        self,
+        logits: torch.Tensor,
+        sampling_metadata: SamplingMetadata,
+    ) -> torch.Tensor:
+        # One idea is implement this as a PyTorch C++ op, and we may
+        # even optimize the logit_bias layout.
+        for i, allowed_token_ids in enumerate(
+                sampling_metadata.has_allowed_token_ids):
+            if allowed_token_ids:
+                logits[i].masked_fill_(
+                    sampling_metadata.allowed_token_ids_mask[i], float("-inf"))
```

Comment:
Can we do a single operation here rather than a loop (make `has_allowed_token_ids` a single bool rather than list).

Reply:
Yeah, just did everything in one masked_fill_